### PR TITLE
WIP: Adding Dragen module

### LINKED ISSUE #4026: new module: dragen
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

## DNA

```[tasklist]
### Tasks
- [x] Provide ped for for [trio](https://github.com/genome-in-a-bottle/giab_data_indexes) @marrip
- [x] Run trio analysis on [HG002, HG003, HG004](https://github.com/genome-in-a-bottle/giab_data_indexes) @xuyangyuio
- [x] Add remaining output files to Dragen module @marrip
```

## RNA

```[tasklist]
### Tasks
- [x] Check if we already have a list of output files @asr081
- [ ] Check if `.gtf` file is present in Dragen references folder @xuyangyuio
- [x] Find RNA samples @marrip
- [ ] Get access to dragen @marrip
- [ ] run test data @marrip
```

## Methylation

```[tasklist]
### Tasks
- [x] Evaluate [test data](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR24827403&display=download) @asr081
- [ ] run test data @marrip
```

## Amplicon

```[tasklist]
### Tasks
```

## Single Cell

```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by marrip**
This is far from finished - still need to figure out test license and add more cases - it should serve as a first draft and base for discussion :wink:

## PR checklist

Closes #4026 
- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/dragen/main.nf [hunk 1]
```diff
@@ -0,0 +1,78 @@
+process DRAGEN {
+    tag "$meta.id"
+    label 'process_long'
+
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/YOUR-TOOL-HERE':
```
Comment (marrip) [inline modules/nf-core/dragen/main.nf:6]: not sure what to do about the container images - there is a docker one but it is not official and I haven't found the Dockerfile for it. One could properly create a docker and singularity image from that if we could get it somehow and put it into the nf-core/biocontainers registry - not sure which one it is.

File: modules/nf-core/dragen/main.nf [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        'registry.hub.docker.com/etycksen/dragen4:4.2.4' }" // somehow the PATH is not containing dragen, when I run it locally it is there
+
+    input:
+    tuple val(meta), path(fastq)
```
Comment (marrip) [inline modules/nf-core/dragen/main.nf:10]: There is several other inputs but I assumed we start with a single use case `--enable-variant-caller true`

File: modules/nf-core/dragen/main.nf [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    tuple val(meta) , path("*.g.vcf.gz.tbi")                        , optional: true , emit: gvcf_gzip_§tbi
+    path "versions.yml"            , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def bin_path = task.ext.bin_path ?: 'dragen'
```
Comment (marrip) [inline modules/nf-core/dragen/main.nf:38]: for some reason, the path to the dragen bin needs to be supplied for the docker image - when I start the container directly I have access to it via `PATH` but if nextflow runs it the `PATH` does not container the right directory `/opt/edico/bin` - anyone an idea why?

File: modules/nf-core/dragen/main.nf [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    def input = ''
+    // from fastq
+    if (fastq) {
```
Comment (marrip) [inline modules/nf-core/dragen/main.nf:42]: Not sure if this is how we want to handle input - we thought that depending on the given files we supply them with the right flags.

File: modules/nf-core/dragen/main.nf [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        if (fastq.size() > 2) {
+            error "Error: cannot have more than 2 fastq files as input."
+        } else {
+            input = '-1 ' + fastq.join(' -2 ')
+        }
+    }
+
+    """
+    ${bin_path}_reset
```
Comment (marrip) [inline modules/nf-core/dragen/main.nf:51]: There is a `dragen_reset` command which should be used prior to a new run - not sure if this goes into a separate module or if we want to run it here? Apparently the reference needs to be loaded as well prior to running things so maybe this needs to be included as well - again, separate module or include it here?

File: modules/nf-core/dragen/main.nf (hunk 2 above)
Comment (marrip) [reply modules/nf-core/dragen/main.nf:10]: the `anable-variant-caller true` requires `--RGID` and `--RGSM`. How should we set them - does the user need to provide something, should they default to `prefix`?

**Comment by marrip** [comment]
just some stuff about AWS:
https://aws.amazon.com/solutions/partners/illumina-dragen/

**Comment by ewels** [comment]
As discussed at the hackathon - let's focus on getting the module as finished first and get it merged using `stub`.

Once we are relatively confident, if we want to continue pushing for CI we can look into it with custom runners with @mashehu and potentially reaching out to illumina for a test license if needed. But let's get the code as finished as possible before we think about this.

File: modules/nf-core/dragen/main.nf (hunk 3 above)
Comment (marrip) [reply modules/nf-core/dragen/main.nf:100]: I found the the file doing this, I believe. In `/etc/profile.d/edico.sh` the path to dragen is added to `PATH` - this is only done for `bash` and not for `sh`. It is probably also linked to having a prompt available or not - that's at least what I think. So nextflow will not have dragen available in `PATH`. Rebuilding the container could potentially solve this by adding it to PATH directly in `Dockerfile`.

**Comment by marrip** [comment]
Yes, I want to try at least, we had some setbacks and also there were other projects that had higher priority.

**Comment by SPPearce** [comment]
> Yes, I want to try at least, we had some setbacks and also there were other projects that had higher priority. Also, I had a couple of questions about how to solve certain parts and never got any reply from the reviewers…

If you want to work on it, give me a shout and I'll review and try and help get it finished off.

**Pull request closed without merging**